### PR TITLE
TZIP-016: Contract metadata

### PR DESCRIPTION
**Pull request by droserasprout**

### CODE REVIEW

File: pytezos/contract/metadata.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+import json
+from os.path import dirname, join
+from typing import Any, Dict, List, Optional, Union
+
+from attr import dataclass
+from jsonschema import validate as jsonschema_validate  # type: ignore
+
+metadata_json_replace_table = {
+    '"return-type":': '"returnType":',
```
Comment (m-kus) [inline pytezos/contract/metadata.py:9]: I would not indulge using kebab case, the current consensus about all metadata (even beyond tzip-16) is that camel case should be used.

File: pytezos/contract/interface.py [hunk 2]
```diff
@@ -200,12 +210,47 @@ def storage(self) -> ContractData:
...
+        metadata_url = self.metadata_url
+        if metadata_url is None:
+            return None
+        logger.info('Trying to fetch contract metadata from `%s`', metadata_url)
+        parsed_url = urlparse(metadata_url)
+        metadata_json = {}
+
+        if parsed_url.scheme in ('http', 'https'):
+            # NOTE: KT1B34qXVRfQrScEaqjjt6cJ5G8LtVFZ7fSc
+            metadata_json = self._get_metadata_json_from_http(metadata_url)
```
Comment (m-kus) [inline pytezos/contract/interface.py:226]: Maybe introduce `ContractMetadata.from_https, ContractMetadata.from_ipfs, ContractMetadata.from_storage` instead? That way one can play with ContractMetadata alone

File: pytezos/contract/metadata.py (hunk 1 above)
Comment (droserasprout) [reply pytezos/contract/metadata.py:9]: This is required by jsonschema shipped with proposal.

File: tests/test_metadata.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    def test_from_json(self):
+        for filename in listdir(self.metadata_path):
+            with self.subTest(filename):
+                with open(join(self.metadata_path, filename)) as file:
+                    metadata_json = json.load(file)
+                    ContractMetadata.from_json(metadata_json)
+
+    def test_run_storage_view(self):
+        contract = PyTezosClient().using('delphinet').contract('KT1RyihALYEsVCcKP7Ya6teCHs9ii5ZHQxvj')
+        view_entrypoint = contract.metadata.call_view('multiply-the-nat-in-storage', contract.code[1])
```
Comment (m-kus) [inline tests/test_metadata.py:21]: Try  `contract.code[1]['args'][0]`

Note that storage section does not necessarily go second.

**Comment by m-kus** [comment]
[TZIP-016 metadata](https://trello.com/c/RKxuU7kU/19-tzip-016-metadata)

**Pull request merged**